### ISSUE TITLE
Allow the order in which preferences are displayed to be specified

### ISSUE THREAD

**Issue by czlee** [EliotBerriot/django-dynamic-preferences#44]
I think (having looked at `registries.py` and the documentation) that there's no current way to specify the order in which preferences are displayed on preference pages. (Please correct me if I'm mistaken.) I think it currently just returns them in the order they come out of the underlying `dict`, which means that it's a function of the keys and possibly not even consistent (or Python implementation-dependent).

There are a couple of ways I could see this working:

1. Return them in the order they were added.
2. Specify a `seq` or `order` class attribute in each `Preference`, which `PreferenceRegistry` will sort by before returning in `preferences()`.

There'd obviously be a (very minor) performance hit here, so there's possibly an argument for making `sorted_preferences` a separate method or `preferences(sort=True)` a keyword argument, so that calls that don't require a sorted list don't need to "wait" for one?

I'll get to this myself eventually, but "eventually" might be a few weeks away so I thought I'd just put this on the radar :smiley:

(Sorry, hit Enter at the wrong time before.)

**Comment by EliotBerriot** [comment]
Thank you for reporting this. You're absolutely right about the varying order. 

Hopefully, I've got a two step solution for this:

- Use [persisting-theory](https://pypi.python.org/pypi/persisting-theory), another package I wrote to handle the registry part, that uses `OrderedDict` instead of `dict`. This way, default ordering will at least be consistent (and match the order in which preferences where registered).
- Update persisting-theory to use (again) [lifter](https://github.com/eliotberriot/lifter) a package I released a few days ago, designed to enable complex filtering/sorting against any python iterable.

The only part that will needs to be done is to find an entry point in `dynamic_preferences` to call the underlying lifter queryset, but you could then do something like:

```python
global_preferences = global_preferences_registry.manager()
# Apply filters on registry objects, not model instances
global_preferences.filter(is_public=True).order_by('label')
```

This may take some times, since lifter's API is not stable right now, but I wrote that package exactly for this kind of use case, and I hope to integrate it here as soon as possible.

Just to be sure I don't miss anything, could you share an example code that shows where you want to sort preferences? I'm not certain of what you mean by "preference pages" (admin page? form builder? something else?).

**Comment by czlee** [comment]
Sorry, loose terminology, by "preferences page" I meant the form returned by the form builder, when inserted into a Django template. The relevant files are here:
https://github.com/czlee/tabbycat/tree/develop/options
`TournamentPreferenceFormView` is in views.py; `tournament_preference_form_builder` is in forms.py. Our app allows you to set up one or more tournaments, and stores preferences for each tournament.

If you're interested in getting a better idea of what we're doing, the easiest way to see this is probably just to deploy an instance of our app, if you have a Heroku account:
[![Deploy](https://www.herokucdn.com/deploy/button.svg)](https://heroku.com/deploy?template=https://github.com/czlee/tabbycat/tree/develop)
Just that link and a Heroku account should do the trick, it should be command-line-free. Hopefully it'll guide you through creating a tournament, you'll need to create a tournament but you won't need to import data or anything to see the options page, which is under "Setup > Configuration" in the left-hand pane. There are ten sections, "Score Rules" through "League Options". @philipbelesky and I work on this project together.

Ideally we'd want to just dictate the order of display. Even within sections there is often a "logical order" in which preferences are best displayed. For example, under "Data Entry" in our app, there's a "Public Password" field and an "Require password for submission" checkbox; they should really be next to each other.

In practice a system in which they were just displayed in the order they were registered would work fine for us, because all our preferences are in the same file. But if we ever split up preferences among different files I wonder what the effect would be?

By the way, since we're here, I'll take a moment to thank you for writing and sharing this module. We had been looking for a good way to manage user configurations for quite a long time, and we were just going to write our own framework until we found yours. It's going very well for us, your app is some really great work.

Let me know if I've missed any information that might be useful.

**Comment by EliotBerriot** [comment]
Thanks for all the details. I don't have an heroku account, but now I understand your use case (that involve the form builder, as I thought).

As a step to fix this, I'm gonna integrate the registry package that use `OrderedDict` internally, so at least order will be consistent each time. It should probably be done by tonight.

**Comment by EliotBerriot** [comment]
This is merged in 0.8, please be careful before upgrading your project, I had to mess up with __init__ and there is a backward incompatible change. Read http://django-dynamic-preferences.readthedocs.org/en/stable/upgrade.html for upgrade instructions.
